spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,136 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
           isTargetRowNotPresent = IsNull(findOutputAttr(joinPlan.output, ROW_FROM_TARGET)),
-          matchedConditions = matchedActions.map(getClauseCondition),
-          matchedOutputs = matchedActions.map(actionOutput(_, target.output)),
+          matchedConditions = matchedConditions,
+          matchedOutputs = matchedOutputs,
           notMatchedConditions = notMatchedActions.map(getClauseCondition),
-          notMatchedOutputs = notMatchedActions.map(actionOutput(_, target.output)),
-          targetOutput = target.output :+ FALSE_LITERAL,
-          deleteOutput = target.output :+ TRUE_LITERAL,
+          notMatchedOutputs = notMatchedActions.map(actionOutput),
+          targetOutput = target.output,
           joinedAttributes = joinPlan.output
         )
         val mergePlan = MergeInto(mergeParams, target, joinPlan)
         val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+
         ReplaceData(target, batchWrite, mergePlan)
     }
   }
 
-  private def actionOutput(clause: MergeAction, targetOutputCols: Seq[Expression]): Seq[Expression] = {
+  private def actionOutput(clause: MergeAction): Option[Seq[Expression]] = {
     clause match {
       case u: UpdateAction =>
-        u.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(u.assignments.map(_.value))
       case _: DeleteAction =>
-        targetOutputCols :+ TRUE_LITERAL
+        None
       case i: InsertAction =>
-        i.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(i.assignments.map(_.value))
     }
   }
 
   private def getClauseCondition(clause: MergeAction): Expression = {
     clause.condition.getOrElse(TRUE_LITERAL)
   }
+
+  private def buildTargetScan(
+      target: DataSourceV2Relation,
+      source: LogicalPlan,
+      cond: Expression): (MergeBuilder, LogicalPlan) = {
+    // Construct the plan to prune target based on join condition between source and target.
+    val mergeBuilder = target.table.asMergeable.newMergeBuilder("merge", newWriteInfo(target.schema))
+    val matchingRowsPlanBuilder = (rel: DataSourceV2ScanRelation) =>
+      Join(source, rel, Inner, Some(cond), JoinHint.NONE)
+    val targetTableScan = buildScanPlan(target.table, target.output, mergeBuilder, cond, matchingRowsPlanBuilder)
+
+    (mergeBuilder, targetTableScan)
+  }
+
+  private def rewriteMatchedActions(
+      matchedActions: Seq[MergeAction],
+      targetOutput: Seq[Expression]): (Seq[Expression], Seq[Option[Seq[Expression]]]) = {
+    val startMatchedConditions = matchedActions.map(getClauseCondition)
+    val catchAllIndex = startMatchedConditions.indexWhere {
+      case Literal(true, BooleanType) =>
+        true
+      case _ =>
+        false
+    }
+
+    val outputs = matchedActions.map(actionOutput)
+    if (catchAllIndex < 0) {
+      // all of the actions have non-trivial conditions. add an action to emit the target row if no action matches
+      (startMatchedConditions :+ TRUE_LITERAL, outputs :+ Some(targetOutput))
```

Comment:
Adding a catch-all case allows us to remove the `targetRowNotPresent` boolean in `MergeIntoExec.applyProjection`. The matched actions now always include a case that matches, so we don't need to check whether no projection matched and there is a target row.